Error should be raised if search binary collection with metric_type: L2

**Please state your issue using the following template and, most importantly, in English.**

**Describe the bug**
```
________________________________________________________________ TestSearchBase.test_search_distance_jaccard_flat_index_L2 ________________________________________________________________

self = <test_search.TestSearchBase object at 0x7ff3811b8e10>, connect = <milvus.client.stub.Milvus object at 0x7ff3813ece80>, binary_collection = 'search_DanKIa3N'

    def test_search_distance_jaccard_flat_index_L2(self, connect, binary_collection):
        '''
        target: search binary_collection, and check the result: distance
        method: compare the return distance value with value computed with L2
        expected: the return distance equals to the computed value
        '''
        nq = 1
        int_vectors, entities, ids = init_binary_data(connect, binary_collection, nb=2)
        query_int_vectors, query_entities, tmp_ids = init_binary_data(connect, binary_collection, nb=1, insert=False)
        distance_0 = jaccard(query_int_vectors[0], int_vectors[0])
        distance_1 = jaccard(query_int_vectors[0], int_vectors[1])
        query, vecs = gen_query_vectors(binary_field_name, query_entities, top_k, nq, metric_type="L2")
        with pytest.raises(Exception) as e:
>           res = connect.search(binary_collection, query)
E           Failed: DID NOT RAISE <class 'Exception'>

entity/test_search.py:634: Failed

```

**Steps/Code to reproduce behavior**
Follow this [guide](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) to craft a minimal bug report. This helps us reproduce the issue you're having and resolve the issue more quickly.

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment details**
master-branch

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/core/src/server/ValidationUtil.cpp
+++ b/core/src/server/ValidationUtil.cpp
@@ -391,6 +391,28 @@ ValidateIndexMetricType(const std::string& metric_type, const std::string& index
 }
 
 Status
+ValidateSearchMetricType(const std::string& metric_type, bool is_binary) {
+    if (is_binary) {
+        // binary
+        if (metric_type == knowhere::Metric::L2 || metric_type == knowhere::Metric::IP) {
+            std::string msg = "Cannot search binary entities with index metric type " + metric_type;
+            LOG_SERVER_ERROR_ << msg;
+            return Status(SERVER_INVALID_ARGUMENT, msg);
+        }
+    } else {
+        // float
+        if (metric_type == knowhere::Metric::HAMMING || metric_type == knowhere::Metric::JACCARD ||
+            metric_type == knowhere::Metric::TANIMOTO) {
+            std::string msg = "Cannot search float entities with index metric type " + metric_type;
+            LOG_SERVER_ERROR_ << msg;
+            return Status(SERVER_INVALID_ARGUMENT, msg);
+        }
+    }
+
+    return Status::OK();
+}
+
+Status
 ValidateSearchTopk(int64_t top_k) {
     if (top_k <= 0 || top_k > QUERY_MAX_TOPK) {
         std::string msg =

--- a/core/src/server/ValidationUtil.h
+++ b/core/src/server/ValidationUtil.h
@@ -45,6 +45,9 @@ extern Status
 ValidateIndexMetricType(const std::string& metric_type, const std::string& index_type);
 
 extern Status
+ValidateSearchMetricType(const std::string& metric_type, bool is_binary);
+
+extern Status
 ValidateSearchTopk(int64_t top_k);
 
 extern Status

--- a/core/src/server/delivery/request/SearchReq.cpp
+++ b/core/src/server/delivery/request/SearchReq.cpp
@@ -79,6 +79,12 @@ SearchReq::OnExecute() {
             if (field->GetFtype() == (int)engine::DataType::VECTOR_FLOAT ||
                 field->GetFtype() == (int)engine::DataType::VECTOR_BINARY) {
                 dimension = field->GetParams()[engine::PARAM_DIMENSION];
+                // validate search metric type and DataType match
+                bool is_binary = (field->GetFtype() == (int)engine::DataType::VECTOR_FLOAT) ? false : true;
+                if (query_ptr_->metric_types.find(field->GetName()) != query_ptr_->metric_types.end()) {
+                    auto metric_type = query_ptr_->metric_types.at(field->GetName());
+                    STATUS_CHECK(ValidateSearchMetricType(metric_type, is_binary));
+                }
             }
         }
 

--- a/tests/milvus_python_test/test_index.py
+++ b/tests/milvus_python_test/test_index.py
@@ -561,8 +561,9 @@ class TestIndexBinary:
         nq = get_nq
         ids = connect.insert(binary_collection, binary_entities)
         connect.create_index(binary_collection, binary_field_name, get_jaccard_index)
-        query, vecs = gen_query_vectors(binary_field_name, binary_entities, top_k, nq)
+        query, vecs = gen_query_vectors(binary_field_name, binary_entities, top_k, nq, metric_type="JACCARD")
         search_param = get_search_param(get_jaccard_index["index_type"], metric_type="JACCARD")
+        logging.getLogger().info(search_param)
         res = connect.search(binary_collection, query, search_params=search_param)
         assert len(res) == nq
 

